module/zfs/arc.c

```diff
@@ -6265,17 +6285,26 @@ arc_read(zio_t *pio, spa_t *spa, const blkptr_t *bp,
 				spa_config_exit(spa, SCL_L2ARC, vd);
 			}
 		} else {
+
 			if (vd != NULL)
 				spa_config_exit(spa, SCL_L2ARC, vd);
+
 			/*
-			 * Skip ARC stat bump for block pointers with
-			 * embedded data. The data are read from the blkptr
-			 * itself via decode_embedded_bp_compressed().
+			 * If the spa has no l2 then it should not contribute to
+			 * l2 miss stats.
 			 */
-			if (l2arc_ndev != 0 && !embedded_bp) {
-				DTRACE_PROBE1(l2arc__miss,
-				    arc_buf_hdr_t *, hdr);
-				ARCSTAT_BUMP(arcstat_l2_misses);
+			if (spa_has_l2) {
```

Comment:
In the current state of this PR if the pool has a cache device but it is faulted (ie `vd == NULL`, what you mention as false-positive in the comment above) we still count it as a miss. I believe this is the correct behavior (as opposed to not counting misses if the device is in a faulted state). Perhaps adding a small comment before this line would be useful.

Reply:
With a little thought I agree that if the cache device has faulted it's okay to count this as a miss.